mipengine/node_registry/node_registry.py

```diff
@@ -0,0 +1,181 @@
+
+        # delete node configuration from kv store
+        self._consul_kv_store.delete(node_id)
+
+    def get_all_nodes(self) -> Dict[str, "NodeInfo"]:
+        all_services = self._consul_agent.services()
+
+        all_nodes = self._parse_nodes_from_services(all_services)
+
+        # pathologies are read from the key value store
+        for node_id in all_nodes.keys():
+            _, data = self._consul_kv_store.get(node_id)
+            if not data:
+                raise self.NodeIDNotInKVStore(node_id)
+            node_params = self._NodeParameters.parse_raw(data["Value"])
+            all_nodes[node_id].pathologies = node_params.pathologies
+
+        return all_nodes
+
+    def _parse_nodes_from_services(
+        self, all_services: Dict[str, str]
+    ) -> Dict[str, "NodeInfo"]:
+        nodes = {}
+        for (service_id, service_info) in all_services.items():
+            tags = service_info["Tags"]
+            role = ""
+            if NodeRole.GLOBALNODE.name in tags:
+                role = NodeRole.GLOBALNODE
+            elif NodeRole.LOCALNODE.name in tags:
+                role = NodeRole.LOCALNODE
+            else:
+                continue
+
+            node_info = self.NodeInfo(
+                ip=service_info["Address"],
+                port=service_info["Port"],
+                role=role,
+            )
+            nodes[service_id] = node_info
+
+        return nodes
+
+    def get_all_global_nodes(self):
+        all_nodes = self.get_all_nodes()
+        return {
+            node_id: node_info
+            for node_id, node_info in all_nodes.items()
+            if node_info.role == NodeRole.GLOBALNODE
+        }
+
+    def get_all_local_nodes(self):
+        all_nodes = self.get_all_nodes()
+        return {
+            node_id: node_info
+            for node_id, node_info in all_nodes.items()
+            if node_info.role == NodeRole.LOCALNODE
+        }
+
+    def get_nodes_with_pathologies(self, pathologies: List[str]):
+        all_local_nodes = self.get_all_local_nodes()
+
+        local_nodes_with_pathologies = {}
+        for node_id, node_info in all_local_nodes.items():
+            if all(p in [p.name for p in node_info.pathologies] for p in pathologies):
+                local_nodes_with_pathologies[node_id] = node_info
+
+        return local_nodes_with_pathologies
+
+    def get_nodes_with_datasets(self, datasets: List[str]):
+        all_local_nodes = self.get_all_local_nodes()
+
+        local_nodes_with_datasets = {}
+        for node_id, node_info in all_local_nodes.items():
+            node_datasets = []
+            for p in node_info.pathologies:
+                node_datasets.extend(p.datasets)
+            if all(d in node_datasets for d in datasets):
+                local_nodes_with_datasets[node_id] = node_info
+        return local_nodes_with_datasets
+
+    def get_db(self, node_id: str) -> "DBInfo":
+        _, data = self._consul_kv_store.get(node_id, index=None)
```

Comment:
As I have implemented it the only binding between each node and its db happens in line 49. It stores the node_id as a key in the key/value store and a db_id as its value. 
So the db_id can be whatever, no convention is enforced. This is why the DB_SERVICE_ID_SUFFIX does not exist in the node_registry.py module but only in the node.py module, because it is not compulsory, you just need to give an arbitrary id for the db.
So you can then call `get_db(node_id)` and get the db service that corresponds to this node, no matter the db_id given to the db service.
My "vision" is in the future that each db process will register itself to the node_registry independently, it will not be the node that also registers the db, apart from itself

With what you propose the binding is string based. From then on we would have to make sure that the db_id has a specific format, {node_id}_db. This would need an extra mechanism to enforce it which would create a strong coupling between the node service and the db service, otherwise it would be very error prone.

And yes this leads to 2 requests to the consul agent, instead of 1 but keeps the services much more decoupled. These requests will happen in the background periodically (every couple of seconds) and are very cheap. We should not sacrifice this loose coupling between node and dbs for the sake of one extra cheap request.